### PR TITLE
crimson/osd/osd_operations/client_request: recover the head and other necessary objects before proceeding

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by xxhdx1985126** [comment]
> /home/jenkins-build/build/workspace/ceph-pull-requests/src/mon/FSCommands.cc:734:15: error: use of undeclared identifier 'fs'
        if (!(fs->mds_map.test_flag(CEPH_MDSMAP_REFUSE_STANDBY_FOR_ANOTHER_FS))) {
              ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/mon/FSCommands.cc:739:17: error: 'mds_map' is a private member of 'Filesystem'
            fs->mds_map.set_flag(CEPH_MDSMAP_REFUSE_STANDBY_FOR_ANOTHER_FS);
                ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/FSMap.h:251:10: note: declared private here
  MDSMap mds_map;
         ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/mon/FSCommands.cc:736:13: error: use of undeclared identifier 'fs'
            fs->fscid,
            ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/mon/FSCommands.cc:746:15: error: use of undeclared identifier 'fs'
          if (fs->mds_map.test_flag(CEPH_MDSMAP_REFUSE_STANDBY_FOR_ANOTHER_FS)) {
              ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/mon/FSCommands.cc:751:19: error: 'mds_map' is a private member of 'Filesystem'
              fs->mds_map.clear_flag(CEPH_MDSMAP_REFUSE_STANDBY_FOR_ANOTHER_FS);
                  ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/FSMap.h:251:10: note: declared private here
  MDSMap mds_map;
         ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/mon/FSCommands.cc:748:15: error: use of undeclared identifier 'fs'
              fs->fscid,

Is the main branch broken?

**Comment by Matan-B** [comment]
> Is the main branch broken?

It was fixed, try rebasing

File: src/crimson/osd/osd_operations/client_request.cc [hunk 1]
```diff
@@ -202,7 +202,14 @@ ClientRequest::process_op(instance_handle_t &ihref, Ref<PG> &pg)
...
     if (pg->is_primary()) {
-      return do_recover_missing(pg, m->get_hobj());
+      auto fut = interruptor::now();
+      auto soid = m->get_hobj();
+      if (!soid.is_head()) {
+        fut = do_recover_missing(pg, soid.get_head());
+      }
+      return fut.then_interruptible([this, pg]() mutable {
+        return do_recover_missing(pg, m->get_hobj());
+      });
```
Comment (liu-chunmei) [inline src/crimson/osd/osd_operations/client_request.cc:215]: could you please to introduce which scenario need recovery head first for clone object?

File: src/crimson/osd/osd_operations/client_request.cc [hunk 2]
```diff
@@ -206,6 +206,9 @@ ClientRequest::process_op(instance_handle_t &ihref, Ref<PG> &pg)
       auto soid = m->get_hobj();
       if (!soid.is_head()) {
         fut = do_recover_missing(pg, soid.get_head());
+      } else if (auto snap = snap_need_to_recover(); snap) {
+        soid.snap = *snap;
+        fut = do_recover_missing(pg, soid);
```
Comment (liu-chunmei) [inline src/crimson/osd/osd_operations/client_request.cc:211]: seems couldn't do rollback here, need find missing the specific snap needed.

Comment (xxhdx1985126) [reply src/crimson/osd/osd_operations/client_request.cc:211]: yeah, I'm working on that.

**Comment by xxhdx1985126** [comment]
@liu-chunmei @Matan-B @athanatos This PR is ready for reviewing, please take a look, thanks.

File: src/crimson/osd/osd_operations/client_request.cc [hunk 3]
```diff
@@ -194,21 +194,56 @@ auto ClientRequest::reply_op_error(const Ref<PG>& pg, int err)
...
+    if (!soid.is_head()) {
+      fut = do_recover_missing(pg, soid.get_head());
+    }
+    return fut.then_interruptible([this, pg, soid] {
+      return pg->obc_loader.with_obc<RWState::RWREAD>(
+        soid.get_head(),
+        [this, pg, soid](auto head, auto) mutable {
+        auto oid = resolve_oid(head->get_head_ss(), soid);
+        assert(oid);
+        return do_recover_missing(pg, *oid
```
Comment (athanatos) [inline src/crimson/osd/osd_operations/client_request.cc:219]: I'd prefer to see this recover(head)->recover(snapshot) pattern moved into a helper.  Maybe rename do_recover_missing to recover_object and make do_recover_missing responsible for this part?

**Review by athanatos — APPROVED** [review]
Stylistic quibble, lgtm otherwise.

File: src/crimson/osd/osd_operations/client_request.cc (hunk 3 above)
Comment (xxhdx1985126) [reply src/crimson/osd/osd_operations/client_request.cc]: Done.

**Comment by Matan-B** [comment]
```
Merge remote-tracking branch 'xxhdx1985126/wip-clone-wait-head' into wip-matanb-crimson-testing-11-22                                    
Merge remote-tracking branch 'matan/wip-matanb-crimson-do_osd_ops_execute-v3' into wip-matanb-crimson-testing-11-22
```
Added #54287 to avoid some known test failures.


https://shaman.ceph.com/builds/ceph/wip-matanb-crimson-testing-11-22/845bb791e48ef302e2ebb3bfb05914db8d667935

**Comment by Matan-B** [comment]
> https://pulpito.ceph.com/matan-2023-11-23_08:05:36-crimson-rados-wip-matanb-crimson-testing-11-22-distro-crimson-smithi/

@xxhdx1985126, results suggest a regression, I'll include #54609 as well - rebased on top of the last successful test tun (to have comparison point). 

Build:
https://shaman.ceph.com/builds/ceph/wip-matanb-crimson-testing-11-19-w54609


Re-based on:
https://pulpito.ceph.com/matan-2023-11-19_17:04:13-crimson-rados-wip-matanb-crimson-testing-11-19-distro-crimson-smithi/

**Comment by Matan-B** [comment]
It looks like one of the 2 PRs introduces an issue spotted by thrashers with snapshot enabled:

https://pulpito.ceph.com/matan-2023-11-23_15:58:34-crimson-rados-wip-matanb-crimson-testing-11-19-w54609-v3-distro-crimson-smithi/

I would suggest running thrash tests locally to try to reproduce the crash.

**Comment by xxhdx1985126** [comment]
> It looks like one of the 2 PRs introduces an issue spotted by thrashers with snapshot enabled:
> 
> https://pulpito.ceph.com/matan-2023-11-23_15:58:34-crimson-rados-wip-matanb-crimson-testing-11-19-w54609-v3-distro-crimson-smithi/
> 
> I would suggest running thrash tests locally to try to reproduce the crash.

Yeah, I'm looking into this. Meanwhile, I submitted a series of PRs, all tagged with 'wip-xuxuehan-testing', fixing snapshot related bugs discovered during my tests. Could you also include those PRs and test it again? Maybe this bug is already fixed. Thanks:-)

**Comment by Matan-B** [comment]
> Yeah, I'm looking into this. Meanwhile, I submitted a series of PRs, all tagged with 'wip-xuxuehan-testing', fixing snapshot related bugs discovered during my tests. Could you also include those PRs and test it again? Maybe this bug is already fixed. Thanks:-)

The tag 'wip-xuxuehan-testing' means that the PR is currently being tested by you. While 'needs-crimson-qa' suggests that the PR awaits testing.
I can also include the other ones, although it will be harder to dissect the cause of the failures. I usually prefer running small batches.

**Comment by xxhdx1985126** [comment]
> > Yeah, I'm looking into this. Meanwhile, I submitted a series of PRs, all tagged with 'wip-xuxuehan-testing', fixing snapshot related bugs discovered during my tests. Could you also include those PRs and test it again? Maybe this bug is already fixed. Thanks:-)
> 
> The tag 'wip-xuxuehan-testing' means that the PR is currently being tested by you. While 'needs-crimson-qa' suggests that the PR awaits testing. I can also include the other ones, although it will be harder to dissect the cause of the failures. I usually prefer running small batches.

Yes, I understand that. But it has been hard for me to run teuthology tests in sepia, due to poor network connections. Right now, I'm running a branch with all the PRs I mentioned above on my own test beds, and it seems to be ok so far.

The reason that I suggested also including those PRs is that there are multiple bugs about snapshoting in the current main branch, so the issue we see in this test may not be a regression. Including thos PRs may clarify whether it's a regression or not.

**Comment by Matan-B** [comment]
> Yes, I understand that. But it has been hard for me to run teuthology tests in sepia, due to poor network connections. Right now, I'm running a branch with all the PRs I mentioned above on my own test beds, and it seems to be ok so far.
> 
> The reason that I suggested also including those PRs is that there are multiple bugs about snapshoting in the current main branch, so the issue we see in this test may not be a regression. Including thos PRs may clarify whether it's a regression or not.

No problem, I'll schedule a full run. By the way, feel free to ping me on Slack when needed next time :)

**Comment by xxhdx1985126** [comment]
> > Yes, I understand that. But it has been hard for me to run teuthology tests in sepia, due to poor network connections. Right now, I'm running a branch with all the PRs I mentioned above on my own test beds, and it seems to be ok so far.
> > The reason that I suggested also including those PRs is that there are multiple bugs about snapshoting in the current main branch, so the issue we see in this test may not be a regression. Including thos PRs may clarify whether it's a regression or not.
> 
> No problem, I'll schedule a full run. By the way, feel free to ping me on Slack when needed next time :)

That's very kind of you, thanks very much:-)

**Comment by Matan-B** [comment]
@xxhdx1985126,
https://pulpito.ceph.com/matan-2023-11-28_14:04:23-crimson-rados-wip-matanb-crimson-testing-11-19-wip-xuxuehan-testing-distro-crimson-smithi/

https://github.com/ceph/ceph-ci/commits/wip-matanb-crimson-testing-11-19-wip-xuxuehan-testing

**Comment by athanatos** [comment]
@xxhdx1985126 @Matan-B As long as it's limited to stuff in crimson/ and doesn't affect other code, I'm fine with merging based on @xxhdx1985126's test runs.  It's the same crimson-rados suite, right?

**Comment by xxhdx1985126** [comment]
> @xxhdx1985126 @Matan-B As long as it's limited to stuff in crimson/ and doesn't affect other code, I'm fine with merging based on @xxhdx1985126's test runs. It's the same crimson-rados suite, right?

Yes. #54761 might need to be approved first, since it's included in the test code.

**Comment by xxhdx1985126** [comment]
@athanatos @Matan-B here's the teuthology test [results.log](https://github.com/ceph/ceph/files/13593632/results.log), all but three are successful, the failures are all due to issues of our testbed.

**Pull request merged**